File: fraiseql-php/src/Attributes/AuthzPolicy.php
<?php

declare(strict_types=1);

namespace FraiseQL\Attributes;

use Attribute;
use FraiseQL\Security\AuthzPolicyType;

/**
 * Marks a type or field with a named, reusable authorization policy.
 *
 * Allows defining policies once and applying them to multiple fields,
 * supporting RBAC, ABAC, custom rules, and hybrid approaches.
 *
 * Example:
 * ```php
 * #[AuthzPolicy(
 *     name: 'piiAccess',
 *     type: AuthzPolicyType::RBAC,
 *     rule: "hasRole(\$context, 'data_manager') OR hasScope(\$context, 'read:pii')",
 *     description: 'Access to Personally Identifiable Information'
 * )]
 * #[GraphQLType(name: 'Customer')]
 * class Customer { }
 * ```
 *
 * @package FraiseQL\Attributes
 */
#[Attribute(Attribute::TARGET_CLASS | Attribute::TARGET_PROPERTY)]
final readonly class AuthzPolicy
{
    /**
     * @param string $name Policy name
     * @param AuthzPolicyType|string $type Policy type (RBAC, ABAC, CUSTOM, HYBRID)
     * @param string $rule Authorization rule expression
     * @param array<string> $attributes Attribute conditions for ABAC
     * @param string $description Policy description
     * @param string $errorMessage Custom error message
     * @param bool $recursive Whether to apply recursively
     * @param string $operations Operation-specific rules
     * @param bool $auditLogging Whether to log access decisions
     * @param bool $cacheable Whether to cache decisions
     * @param int $cacheDurationSeconds Cache duration in seconds
     */
    public function __construct(
        public string $name,
        public AuthzPolicyType|string $type = AuthzPolicyType::CUSTOM,
        public string $rule = '',
        public array $attributes = [],
        public string $description = '',
        public string $errorMessage = '',
        public bool $recursive = false,
        public string $operations = '',
        public bool $auditLogging = false,
        public bool $cacheable = true,
        public int $cacheDurationSeconds = 300,
    ) {
    }
}

File: fraiseql-java/src/main/java/com/fraiseql/core/AuthzPolicy.java
package com.fraiseql.core;

import java.lang.annotation.*;

/**
 * Defines a reusable authorization policy that can be referenced by other decorators.
 * Policies centralize authorization logic for consistency and maintenance.
 *
 * Usage:
 * <pre>
 * // Define a policy
 * @AuthzPolicy(
 *     name = "piiAccess",
 *     description = "Access to Personally Identifiable Information",
 *     rule = "hasRole($context, 'data_manager') OR hasScope($context, 'read:pii')"
 * )
 * public class PIIAccessPolicy {}
 *
 * // Use the policy
 * @GraphQLType
 * public class Customer {
 *     @GraphQLField
 *     public String id;
 *
 *     @GraphQLField
 *     @Authorize(policy = "piiAccess")
 *     public String email;
 *
 *     @GraphQLField
 *     @Authorize(policy = "piiAccess")
 *     public String phoneNumber;
 * }
 *
 * // Policy with attributes
 * @AuthzPolicy(
 *     name = "financialData",
 *     type = AuthzPolicyType.ATTRIBUTE,
 *     attributes = {"clearance_level >= 3", "department == 'finance'"},
 *     description = "Access financial records requires clearance level 3 and finance department"
 * )
 * public class FinancialDataPolicy {}
 *
 * // Policy with multiple conditions
 * @AuthzPolicy(
 *     name = "auditAccess",
 *     type = AuthzPolicyType.HYBRID,
 *     rule = "hasRole($context, 'auditor')",
 *     attributes = {"audit_enabled == true"},
 *     description = "Auditors can view audit trails when audit is enabled"
 * )
 * public class AuditAccessPolicy {}
 * </pre>
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.TYPE)
public @interface AuthzPolicy {
    /**
     * Unique name for this policy.
     * Used when referencing the policy via @Authorize(policy = "name")
     *
     * Example: "piiAccess", "financialData", "adminOnly"
     */
    String name();

    /**
     * Description of what this policy protects and why.
     *
     * Example: "Restricts access to personally identifiable information"
     */
    String description() default "";

    /**
     * Custom authorization rule expression.
     * Used for role-based and custom logic policies.
     *
     * Example: "hasRole($context, 'admin') OR isOwner($context.userId, $field.ownerId)"
     */
    String rule() default "";

    /**
     * List of attribute conditions to check.
     * Used for attribute-based access control (ABAC).
     *
     * Examples:
     * - "clearance_level >= 3"
     * - "department == 'finance'"
     * - "country IN ('US', 'CA', 'MX')"
     */
    String[] attributes() default {};

    /**
     * Type of authorization policy.
     * - RBAC: Role-based access control
     * - ABAC: Attribute-based access control
     * - CUSTOM: Custom rule expressions
     * - HYBRID: Combination of multiple approaches
     */
    AuthzPolicyType type() default AuthzPolicyType.CUSTOM;

    /**
     * Whether to cache authorization decisions for this policy.
     */
    boolean cacheable() default true;

    /**
     * Cache duration in seconds.
     */
    int cacheDurationSeconds() default 300;

    /**
     * Whether this policy applies hierarchically to child fields.
     */
    boolean recursive() default false;

    /**
     * Operations this policy applies to.
     * If empty, applies to all operations.
     *
     * Example: "read", or "create,update,delete"
     */
    String operations() default "";

    /**
     * Audit logging configuration.
     * Whether to log access attempts for this policy.
     */
    boolean auditLogging() default true;

    /**
     * Error message to return when policy check fails.
     */
    String errorMessage() default "";

    /**
     * Enum for policy types.
     */
    enum AuthzPolicyType {
        /**
         * Role-based access control (RBAC).
         */
        RBAC,

        /**
         * Attribute-based access control (ABAC).
         */
        ABAC,

        /**
         * Custom rule expressions.
         */
        CUSTOM,

        /**
         * Hybrid approach combining multiple methods.
         */
        HYBRID
    }
}
